libraries/AP_Common/ExpandingString.cpp

```diff
@@ -0,0 +1,102 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  expanding string, used for @SYS filesystem output
+ */
+
+#include "ExpandingString.h"
+
+extern const AP_HAL::HAL& hal;
+
+#define EXPAND_INCREMENT 512
+
+/*
+  expand the string buffer
+ */
+bool ExpandingString::expand(uint32_t min_needed)
+{
+    // expand a reasonable amount
+    uint32_t newsize = (5*buflen/4) + EXPAND_INCREMENT;
+    if (newsize - used < min_needed) {
+        newsize = used + min_needed;
+    }
+    
+    // add one to ensure we are always null terminated
+    void *newbuf = hal.util->std_realloc(buf, newsize+1);
+
+    if (newbuf == nullptr) {
+        allocation_failed = true;
+        return false;
+    }
+
+    buflen = newsize;
+    buf = (char *)newbuf;
+
+    return true;
+}
+
+/*
+  print into the buffer, expanding if needed
+ */
+void ExpandingString::printf(const char *format, ...)
+{
+    if (allocation_failed) {
+        return;
+    }
+    if (buflen == used && !expand(0)) {
+        return;
+    }
+    int n;
+
+    /*
+      print into the buffer, expanding the buffer if needed
+     */
+    while (true) {
```

Comment:
I'm finding this construct unnecessarily complex/dangerous.  Just have two invocations of `vsnprintf``?

Reply:
i'd really rather not double the format processing time. Stare at it a bit longer and I think you'll be OK with it